### ISSUE TITLE
Is there a proper way to determine the correct ordering of the images (Image Index)

### ISSUE THREAD

**Issue by MartinOrtiz** [fo-dicom/fo-dicom#1175]
Say I have a folder of *.dcm files
000000.dcm
000001.dcm
000002.dcm
000003.dcm
000004.dcm

In most cases, I can just process the files in order, because the image index will match the file order.
000000.dcm (Image Index = 0)
000001.dcm (Image Index = 1)
000002.dcm (Image Index = 2)
000003.dcm (Image Index = 3)
000004.dcm (Image Index = 4)

But in other cases, upon examining the Dicom data items....it will be
000000.dcm (Image Index = 4)
000001.dcm (Image Index = 3)
000002.dcm (Image Index = 1)
000003.dcm (Image Index = 0)
000004.dcm (Image Index = 2)

That is the image index does not match the file ordering.

So then I realized after debugging, that I need to get the DicomItem with the Image Index tag,
DicomItem ImageIndex = dicom.Dataset.GetDicomItem<DicomItem>(DicomTag.ImageIndex);
and get the value of the image's Index
dicom.Dataset.TryGetSingleValue(DicomTag.ImageIndex, out ImageIndexVal);

And use that as the correct Image Index, and it worked.....

But then I ran into *.dcm folders examples, where the ImageIndex was not defined...
They set and used the InstanceNumber for the Image's Index

So I did the following, and it worked (for this folder of dcm files)

DicomItem InstanceNumber = dicom.Dataset.GetDicomItem<DicomItem>(DicomTag.InstanceNumber);
dicom.Dataset.TryGetSingleValue(DicomTag.InstanceNumber, out InstanceNumberVal);

So my question is this, using your library....
Is there a way to determine what strategy I should use to correctly asign the images index for all the *.dcm files in the folder?

In other words...
When can I know to just use the file order in the folder, and use that to assign my image indexes?
When can I know that I need to use the tagged ImageIndex DicomItem's value for my image indexes?
When can I know that I need to use the tagged InstanceNumber DicomItem's value for my image indexes?

Is there an API for indicating what I must use in assigning my Image Indexes?

**Comment by EdwardMarcus** [comment]
With different folder of images I sort as follows

1) study date
2) series number
3) instance number

When study A precedes study B :

A - 1 - 1   (Study A - Series 1 - Instance 1)
A - 1 - 2   (Study A - Series 1 - Instance 2)
A - 2 - 1   (Study A - Series 2 - Instance 1)
B - 1 - 1   (Study B - Series 1 - Instance 1)


I work with echo cardiograms, and most of the time this info is reliably embedded. Sometimes series number is zero, but I have only seen this when there is one series in the study- so its not an issue.

**Comment by MartinOrtiz** [comment]
I am creating 3D volumes

![ChestStudy](https://user-images.githubusercontent.com/11322523/117058815-14bba300-aced-11eb-98d1-b426ed263743.png)

**Comment by MartinOrtiz** [comment]
Here is a skull 3D volume

![skullstudy](https://user-images.githubusercontent.com/11322523/117059607-fc985380-aced-11eb-97bc-8d46da7eeda7.png)

**Comment by EdwardMarcus** [comment]
Maybe slice location attribute would order a stack of 2d slices.

**Comment by MartinOrtiz** [comment]
I download DiCOM samples from
[https://imaging.nci.nih.gov/nbia-search/](url)

What I do, is grab one folder from the downloads (there may be multiple studies)

In order to correctly build my 3D volume, I must (of course) process and "stack" the images in the correct order.
Well, the question then is, how is this determined?

**Comment by EdwardMarcus** [comment]
Can you dump one of the files into a list and send?

**Comment by MartinOrtiz** [comment]
> 
> Maybe slice location attribute would order a stack of 2d slices. 

I have looked at that as a possible aid....

I have found in the examples I download, that there is no consensus....
In some cases I can use the file order. 
So that file01 -> Image Index = 1
So that file02 -> Image Index = 2
So that file0n -> Image Index = n
In other cases, this is not true, the file order and index order does not match, thus I must use some other method.
In some cases I can use the ImageIndex tag value 
So that ImageIndex tag = 1 -> actual Image Index = 1
So that ImageIndex tag = 2 -> actual Image Index = 2
So that ImageIndex tag = n -> actual Image Index = n

BUT in some cases there is no ImageIndex tag, and then I must use the Instance Number in assigning Image Index values
So that Instance Number tag = 1 -> actual Image Index = 1
So that Instance Number tag = 2 -> actual Image Index = 2
So that Instance Number tag = n -> actual Image Index = n

**Comment by MartinOrtiz** [comment]
Is there no definitive way other than examining the DicomItems properties (Ie in debugger)

**Comment by EdwardMarcus** [comment]
What is the modality – MRI , X ray. Etc..

**Comment by MartinOrtiz** [comment]
Would knowing that help determine what to do?
Right now, I only parse/extract the image files, I don't deal with or need to know other stuff.
I am prototyping a 3D volume viewer, so am focusing just on 3D volume building and display.

**Comment by MartinOrtiz** [comment]
This is not the whole folder, it's too large, just the first 20 *.dcm files.

[2-.625 mm Chest-16577.zip](https://github.com/fo-dicom/fo-dicom/files/6423655/2-.625.mm.Chest-16577.zip)

**Comment by EdwardMarcus** [comment]
You want to sort by slice location

0020:1041 - Slice Location

**Comment by MartinOrtiz** [comment]
I will examine that...

**Comment by MartinOrtiz** [comment]
> 
> Maybe slice location attribute would order a stack of 2d slices. 

This works. I ordered my image indexes based on the Slice location.

Now, there's probably no answer to this question...
If you have a directory of *.dcm files, why are all the slices in random order? 
(in some cases they are in order, in some cases not)
It would seem to be a logical process where the file list order and slice order would be the same...(and thus the Image Index order)

**Comment by MartinOrtiz** [comment]
Now should I stick with Slice Location? Is this the "standard" process for determining and assigning of Image Indexes?
(thanks for help!)

**Comment by EdwardMarcus** [comment]
Glad it worked out.

You can take a look at the filename - the records in the DICOM directory might be alphabetically ordered by the filename.

I am assuming slice location is a required element for the modality of these images- if so you can reliably count on it to be in your datasets.

**Comment by chen-wu** [comment]
The only reliable way is to reorder images by "Image Position Patient". Do not use other attributes, especially your "Slice Location".
It is a type 3 attribute, and just a relative position.

**Comment by EdwardMarcus** [comment]
Thank you for this insight

into the correct method from the standard.

**Comment by MartinOrtiz** [comment]
I am not sure what the issue is....

If I do this...ImagePositionPatient  is NOT null
DicomItem ImagePositionPatient = dicom.Dataset.GetDicomItem<DicomItem>(DicomTag.ImagePositionPatient);
If I add ImagePositionPatient variable to watch...it does have a StringValue of "198.146\\151.316\\70.4268"


But if I do this.....Try to get the the value of it in my code....
b = false, and ImagePositionPatientVal  is not assigned any value.
string ImagePositionPatientVal = "";
bool b = dicom.Dataset.TryGetSingleValue(DicomTag.ImagePositionPatient, out ImagePositionPatientVal);

Like I said, ImagePositionPatient,StringValue shows up in debugger watch window.
But ImagePositionPatient,StringValue does not appear to be an accessible property in my code.

So I can not get it's value in my code, even though it is displayed in my watch window.


ImagePositionPatient.StringValue

**Comment by EdwardMarcus** [comment]
The image position data is multi-valued.

**Comment by MartinOrtiz** [comment]
```
`                string ImagePositionPatientVal = "";

                DicomItem ImagePositionPatient = dicom.Dataset.GetDicomItem<DicomItem>(DicomTag.ImagePositionPatient);

                bool b = dicom.Dataset.TryGetSingleValue(DicomTag.ImagePositionPatient, out ImagePositionPatientVal);
`
```

in above code snippet
ImagePositionPatient = not null
ImagePositionPatientVal = ""  (not assigned a value)

file attached...
[000000.zip](https://github.com/fo-dicom/fo-dicom/files/6444129/000000.zip)

**Comment by MartinOrtiz** [comment]
> 
> The image position data is multi-valued. 

I tried this instead, and got the values....which one should I use?
                string[] values;
                bool b = dicom.Dataset.TryGetValues(DicomTag.ImagePositionPatient, out values);

Do I use the Position value?

**Comment by MartinOrtiz** [comment]
Yes, 3 values.

**Comment by EdwardMarcus** [comment]
Great - You are good to go.

**Comment by chen-wu** [comment]
You can check the standard: [Image Plane Module](http://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.7.6.2.html#sect_C.7.6.2.1.1)
As you want to build volume from a series of images, it would be better to understand this patient-based coordinate system. It will help you to build the volume in correct order, even if the patient is scanned in different orientation.

**Comment by MartinOrtiz** [comment]
What do I do when a DICOM file contains all the images in one file? 
Some DICOM files bundle all the images into the one file, instead of a separate DICOM file per image (with all the tag information also per image).
I have a DICOM file, I exported all the images in frame order, as that was the only ordering chose I could see to use.
Some bundled/single file DICOM files seem to be frame ordered properly. That is, I can base my image indexes on the frame ordering.
Sometimes this works, sometimes not...that is, some DICOM files seem to be partially ordered.
For example, I exported all the images out of one file, using the frame ordering as the index ordering.
When I built my 3D volume from those images, One was a skull, the skull had the jaw on top, the face in the middle and the skull cap on bottom. So it appears that the ordering from the original DICOM file was not 100% accurate.

What do I do in this case?

**Comment by EdwardMarcus** [comment]
Maybe you can check if this attribute is specified

DICOM has an attribute called Spacing Between Slices (0018, 0088) that gives the distance between two adjacent slices (perpendicular to the image plane)


In some case the spacing might positive, in others negative.

**Comment by MartinOrtiz** [comment]
As I said....this is one single DICOM file, not a folder of DICOM files....
It's one DICOM file, that has all the images embedded into the one file.
As far as I can see, the individual images do not have their own tags to provide information, the only tags are with the "main" DICOM file itself. So I can not examine tags per image, since there are none.
I hope that makes sense.

For a folder of DICOM files, each DICOM file has only one image, and there are tags specifying a lot of information pertaining to that one image (width, height, slice order....etc....etc...)

But in the case where a single DICOM file contains multiple images, are there also multiple tags pertaining to each image? if so, where is it? I only see one set of tags, pertaining to the main file itself.

I hope my explanation makes sense.....

**Comment by MartinOrtiz** [comment]
In other words, when I have a folder of DICOM files, than each DICOM file which only has one image in it, will contain enough tags and information for me to determine the indexes for the images, so that I can create a correct 3D volume out of them.

But when there is only one DICOM, that has multiple images, it does not contain any tag information on a per image basis, so that I can determine how to do the correct ordering (other than using the Frame order).

**Comment by MartinOrtiz** [comment]
Yes, in most cases this is true, and everything is ordered properly and there is no issue with the 3D volume creation.

But in some cases it does not appear to be the case, I guess, there is nothing that can be done about it, when this is the case.

**Comment by chen-wu** [comment]
Please read the DICOM Standard. When you try to read a DICOM file, the standard is always the best manual. 
Just check the SOP Class UID, and find the that part in [PS3.3](http://dicom.nema.org/medical/dicom/current/output/html/part03.html)

The DICOM file you descript is called "Enhanced", the structure is well defined in DICOM standard. Usually a multi-frame DICOM contains a per-frame sequence. Which contains some unique attributes for each frame. Other attributes are shared across all frames. 
So you can always assign such information to each frame, and reorder them. 

A "Enhanced" DICOM just packages some frames together in one object. The order of frames may or may not be meaningful. You can even package several sequences of images into one enhanced file. 

Inside your application, you should split the frames, assign the per-frame information, reorder frames to several image stacks, then display them by a fixed Image Orientation Patient.

Or, just ignore the SOP Classes you cannot handle.

**Comment by MartinOrtiz** [comment]
This is how I process enhanced dcm files...
I can see a "SOP Class UID" in dataset in debugger watch window, but what do I do with it?
I don't see where the meta data is for each image....
(below exports the images in frame order)

`        public void ParseDCMFile(string SourcePath, string DestPath, string BaseName, int dim)
        {
            DicomFile dicom = DicomFile.Open(SourcePath);
            var images = new DicomImage(dicom.Dataset);

            for (var count = 0; count < images.NumberOfFrames; count++)
            {
                var image = images.RenderImage(count).As<Bitmap>();
                var resizedImage = new Bitmap(image, new System.Drawing.Size(dim, dim));
                string n = (count.ToString().PadLeft(4, '0'));
                resizedImage.Save(DestPath + @"\" + BaseName + n + ".png", System.Drawing.Imaging.ImageFormat.Png);
            }
        }`
